### PR TITLE
Refactor social AUTH modules

### PR DESCRIPTION
**Pull request by hishamco**
After @sebastienros comments on the last standup meeting, there's no need to have interface for social AUTH settings services, this PR make everything consistent plus abstract almost common implementation details

### CODE REVIEW

**Comment by Piedone** [comment]
I won't be able to review this for quite a while, so if anybody else can, please do.

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I'm OK with the `OAuthSettingsService<TAuthenticationSettings>` base class, but why is it good to remove the service interfaces? We lose the ability to override them (for extensibility and testing).

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: I don't think these new line breaks are needed. Same in `GoogleAuthenticationSettings`, `AzureADSettings`, `MicrosoftAccountSettings`, `MicrosoftAccountSettings`.

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Security/Services/OAuthSettingsService.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Security/Services/OAuthSettingsService.cs:43]:
```suggestion
        oldSettings.Put(nameof(TAuthenticationSettings), settings);
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: It's good for clarity & readability

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: > We lose the ability to override them (for extensibility and testing).

Totally agree, I just react toSeb feedback on the meeting, but I think it's fine to leave it as it is

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: @sebastienros please elaborate, then.

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: There's a reason we don't do this in every single place though :).

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: May be because each service has default implementation, but still I'm agree with you :)

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: May be because each service has default implementation, but still I'm agree with you :)

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: That's why I surrprised in many places in the code base ;)

**Comment by Piedone** [comment]
Any news here, @hishamco?

**Comment by hishamco** [comment]
@Piedone I need to revise all the pending PRs, seems it's the time make my PRs ready, thanks for reminder

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I still don't think we should remove the service interfaces. It's a step back.

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: I still think these line breaks are not needed. It doesn't add any readability any more than adding line breaks between any two lines of code.

**Review by Piedone — CHANGES_REQUESTED** [review]

**Comment by hishamco** [comment]
Adding lines between properties and methods is a good practice I have seen. Even ASP.NET Core team use it in the code base

**Comment by Piedone** [comment]
Please don't move threaded convos out because it makes things harder to follow.

**Comment by hishamco** [comment]
Believe it or not, I do it because I can't reply to your comment above :(

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: Adding lines between properties and methods is a good practice I have seen. Even ASP.NET Core team uses it in the code base

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I'm not sure why we need if we have an abstract class?!!

**Comment by hishamco** [comment]
GitHub surpised me, seems we you comment on the code it duplicate it

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: If the properties have anything else on them, like attributes, or documentation, or they are multi-line, then sure. But otherwise, why? Why don't we add line breaks between private fields of a class too?

**Comment by hishamco** [comment]
1st

Screenshot 2024-03-14 021426

2nd

Screenshot 2024-03-14 021441

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: The same reason as we do it for services generally: loose coupling. However, these classes aren't even abstract. All are concrete types that are injected, removing the option to override them.

File: src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Settings/GithubAuthenticationSettings.cs:9]: That's the code lines I know from years :)

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I didn't get you, I'm not sure why we should keep interfaces with the same methods while we can introduce a new abstract class to do the same thing

**Comment by Piedone** [comment]
We talked about this under the ACS PR. This is how GH's PR comments work for some reason. It's not good, and I've actually written their support about it (weeks ago, no reply yet) but we should still scroll up and reply to the threads.

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I will bring the interfaces again

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: Zoltan while we have many interfaces, I think we need only one for all services that could be implemented in the abstract class, right?

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: Not really. If I want to override `FacebookService`, I need a specific interface for that, i.e. `IFacebookService` as it is in `main`. A common interface won't work for this.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: Could you please let me know how all the interfaces will interact with the newly added abstract class?

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: They don't need to, just e.g. `FacebookService` needs to implement `IFacebookService`, as it was. All these interfaces can remain as they were, and since the abstract class implements them, in e.g. `FacebookService` you don't need to add anything else.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: That means nothing to be removed except a few types have been added

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: Well, up to you how to go forward. The interfaces should stay because they have a use, they can't just be removed, as it will break people's code.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: Return all interfaces will make things harder because the abstract class should implement all the auth interfaces which do not make sense, all are equivalent except the settings type. If we introduce a generic type we will find ourselves using a single interface instead which is why I moved the implementation into an abstract class

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I can only repeat myself. Refactoring that breaks all consumers of these authentication services is not suitable. We have more than 400 interfaces in OC for such services, for a good reason. I really don't want to debate further why we should keep them. By looking at the code it seems to me that the only thing you need to do is re-add the interfaces and revert the dependency registrations, nothing else to change. It's something really simple.

Without the interfaces, I don't approve this PR. It's your call whether you want to rework the PR to keep them, or close it.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: As I said before my PR is based on @sebastienros suggestion nothing else. Moreover, I never mind reverting all the removed interfaces but the thing that I still do not get is how the interfaces will be used in the introduced abstract class

If you think this will take so long, please close the PR, I don't want to continue in something I wasn't convinced with

Thanks

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I suggest you just re-add the interfaces to see in action that nothing else with the base class needs to change.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookService.cs:8]: I just reverted the interfaces back as you said but again a generic interface make sense to me rather than have 4 - 6 interfaces with the same methods

File: src/OrchardCore.Modules/OrchardCore.GitHub/Recipes/GithubAuthenticationSettingsStep.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.GitHub/Recipes/GithubAuthenticationSettingsStep.cs:30]: `GetSettingsAsync()` returns an immutable settings object that won't be possible to update in the next lines, so this should stay Load*. Same for all the *Step classes.

**Review by Piedone — CHANGES_REQUESTED** [review]
Please test all logins, as well as imports, if you haven't done already.

**Comment by sebastienros** [comment]
Closing as stale and not much value. There is also a new settings helper that is now used in these classes.

**Pull request closed without merging**
